bip-0341.mediawiki

```diff
@@ -0,0 +1,311 @@
+If the taproot output key is an aggregate of keys, there is the possibility for a malicious party to add a script path without being noticed by the other parties.
+This allows to bypass the multiparty policy and to steal the coin.
+MuSig key aggregation does not have this issue because it already causes the internal key to be randomized.
+
+The attack works as follows: Assume Alice and Mallory want to aggregate their keys into a taproot output key without a script path.
+In order to prevent key cancellation and related attacks they use [https://eprint.iacr.org/2018/483.pdf MSDL-pop] instead of MuSig.
+The MSDL-pop protocol requires all parties to provide a proof of possession of their corresponding secret key and the aggregated key is just the sum of the individual keys.
+After Mallory receives Alice's key ''A'', Mallory creates ''M = M<sub>0</sub> + int(t)G'' where ''M<sub>0</sub>'' is Mallory's original key and ''t'' allows a script path spend with internal key ''P = A + M<sub>0</sub>'' and a script that only contains Mallory's key.
+Mallory sends a proof of possession of ''M'' to Alice and both parties compute output key ''Q = A + M = P + int(t)G''.
+Alice will not be able to notice the script path, but Mallory can unilaterally spend any coin with output key ''Q''.
+</ref>
+* The remaining scripts should be organized into the leaves of a binary tree. This can be a balanced tree if each of the conditions these scripts correspond to are equally likely. If probabilities for each condition are known, consider constructing the tree as a Huffman tree.
+
+'''Computing the output script''' Once the spending conditions are split into an internal key <code>internal_pubkey</code> and a binary tree whose leaves are (leaf_version, script) tuples, the output script can be computed using the Python3 algorithms below. These algorithms take advantage of helper functions from the [bip-0340/referency.py BIP340 reference code] for integer conversion, point multiplication, and tagged hashes.
+
+First, we define <code>taproot_tweak_pubkey</code> for 32-byte [bip-0340.mediawiki#design BIP340] public key arrays.
+In addition to the tweaked public key byte array, the function returns a boolean indicating whether the public key represents the tweaked point or its negation.
+This will be required for spending the output with a script path.
+In order to allow spending with the key path, we define <code>taproot_tweak_seckey</code> to compute the secret key for a tweaked public key.
+For any byte string <code>h</code> it holds that <code>taproot_tweak_pubkey(pubkey_gen(seckey), h)[0] == pubkey_gen(taproot_tweak_seckey(seckey, h))</code>.
+
+<source lang="python">
+def taproot_tweak_pubkey(pubkey, h):
+    t = int_from_bytes(tagged_hash("TapTweak", pubkey + h))
+    if t >= SECP256K1_ORDER:
+        raise ValueError
```

Comment:
@sipa I understand it's an extremely rare occurrence, but shouldn't there be some sort of consideration to prevent failure of this nature. (ie. RFC6979 has the last step loop if the resulting nonce doesn't produce valid r and s values)

Perhaps we can say something like:

```python
    t = int_from_bytes(tagged_hash("TapTweak", pubkey + h))
    while t >= SECP256K1_ORDER:
        t = int_from_bytes(tagged_hash("TapTweak", bytes_from_int(t)))
```

etc.

Or would the proper solution be "generate a new pubkey or tweak"?

Reply:
if it does happen than you've successfully broken SHA256 :)
it's like "guessing" Satoshi's private keys.

Reply:
arghh meant to edit not delete,
I originally wrote that this is somewhat equivalent to breaking SHA256 or guessing Satoshi's private keys(probably easier). but then I realized that in schnorr signing we do mod order to "solve this". so I'm not sure why the difference @sipa ?

Reply:
We don't care about this condition. The probability of this overflow occurring is 1 in 2^127.65 - similar to our security level (the number of operations an attacker is expected to need to break ECDLP), so we can treat it as effectively impossible.

My view is that a specification shouldn't be burdened by addressing impossible events, or at least, it should specify the easiest way of dealing with it that remains correct. For nonce creation we use a modulo operation, as that is likely how you'd naively implement it if you weren't aware this overflow was possible at all. For the tweak here, the spec mimics the behavior of BIP32 additive key tweaking.

Reply:
(and e)

Reply:
ok, sounds reasonable.